dbms/src/Processors/Executors/PipelineExecutor.cpp

```diff
@@ -459,49 +501,31 @@ void PipelineExecutor::executeSingleThread(size_t thread_num, size_t num_threads
 #if !defined(__APPLE__) && !defined(__FreeBSD__)
     /// Specify CPU core for thread if can.
     /// It may reduce the number of context swithches.
-    cpu_set_t cpu_set;
-    CPU_ZERO(&cpu_set);
-    CPU_SET(thread_num, &cpu_set);
-    if (sched_setaffinity(0, sizeof(cpu_set_t), &cpu_set) == -1)
-        LOG_TRACE(log, "Cannot set affinity for thread " << num_threads);
+    /*
```

Comment:
Recently I came across this issues, where all queries used only one thread, and hence all of them executed only on the first core, and even though this has `num_threads > 1` this still does not solve the following issues:
- it does not reset the affinity when the thread returned to the thread pool, and this means that this affinity will be preserved until the restart (and in my case with `experimental_use_processors=1` almost every CH thread was executed only on first core because of this after executing X queries)
- there can be more `num_threads` then the cores available

I am curious about this, since even though this is commented right now, I guess it can be uncommented anytime.

So @KochetovNicolai what do you think about making this under setting only? (and I guess that some core allocator should be introduced for load balancing)

> `/// It may reduce the number of context swithches`

P.S. do you have any numbers on how this can improve things?

Reply:
You are right, it is needed to be implemented in a different way before enabling.
I just had a query which was about 10% faster with the affinity enabled, and `perf stat` showed less cache misses. Also, `top` showed more evenly distributed cpu loading for new pipeline comparing with the old one. So, my hypothesis was that too often context switches could lead to cash misses.

I enabled affinity to compare perftest results. But it showed no improvement. On the contrary, it slowed down queries in average. So, I have commented this code by now. And will most probably remove it later.